File: 01/1.6/sqrt-iter.hs
sqrtIter :: Float -> Float -> Float
sqrtIter guess x =
  if goodEnough guess x 0.00001 then guess else sqrtIter (improve guess x) x

goodEnough :: Float -> Float -> Float -> Bool
goodEnough guess x accuracy = abs (square guess - x) < accuracy

square :: Float -> Float
square x = x * x

improve :: Float -> Float -> Float
improve guess x = (guess + x / guess) / 2

sqrt' :: Float -> Float
sqrt' = sqrtIter 1.0

main :: IO ()
main = do
  let res = sqrt' 16.0
  putStrLn ("result is " ++ show res)

File: 01/1.6/sqrt-iter.ex
defmodule SqrtIter do
  @spec sqrtIter(number(), number()) :: number()
  def sqrtIter(guess, x) do
    if goodEnough?(guess, x, 0.00001) do
      guess
    else
      sqrtIter(improve(guess, x), x)
    end
  end

  @spec goodEnough?(number(), number(), number()) :: boolean()
  def goodEnough?(guess, x, accuracy) do
    abs(square(guess) - x) < accuracy
  end

  @spec square(number()) :: number()
  def square(x), do: x*x

  @spec improve(number(), number()) :: number()
  def improve(guess, x), do: (guess + x / guess) / 2

  @spec sqrt(number()) :: number()
  def sqrt(x), do: sqrtIter(1.0, x)
end
